companies/views.py

```diff
@@ -1,86 +1,83 @@
 
-    def post(self, request, *args, **kwargs):
-
-        auth_response = JWT_authenticator.authenticate(request)
-        if auth_response is None:
-            return Response({'error': "Invalid token"}, status=status.HTTP_401_UNAUTHORIZED)
-        user_id = auth_response[1].get('user_id')
-        request.data['investor'] = user_id
-        serializer = SubscriptionSerializer(data=request.data)
+    def post(self, request):
+        serializer = CompaniesSerializer(data=request.data)
         if serializer.is_valid():
-            subs = Subscription.objects.filter(investor=user_id, company=request.data.get('company'))
-            if not subs:
-                subscription = serializer.save()
-            else:
-                subscription = subs.first()
-            company = subscription.company
-            brand = company.brand
-            subscription_id = subscription.subscription_id
-            message = f"You're subscribed to {brand}, subscription id: {subscription_id}"
-            return Response({'message': message}, status=status.HTTP_201_CREATED)
-        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+            print(serializer)
+            company = Company(**serializer.data)
+            company.save()
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+        else:
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+
+class CompaniesRetrieveUpdateView(APIView):
+    permission_classes = (IsAuthenticated)
+    def get(request, pk=None):
+        if pk:
+            company = Company.get_company(company_id=pk)
+            return Response(company.get_info(), status=status.HTTP_200_OK)
+        type = request.data.get('company_type')
+        if not type:
+           companies = Company.get_all_companies_info()
+        else:
+            companies = Company.get_all_companies_info(type)
+        return Response(companies, status=status.HTTP_200_OK)
+    
+class SubscriptionCreateAPIView(APIView):
+    permission_classes = (IsAuthenticated, IsRelatedToCompany, IsInvestor)
 
+    def post(self, request):
+        profile_id = request.user.relation_id
+        company_id = request.data.get('company_id')
+        try:
+            check_sub = Subscription.get_subscription(
+                investor=profile_id, company=company_id)
+        except Subscription.DoesNotExist:
+            try:
+                data = {'investor': CompanyAndUserRelation.get_relation(relation_id=profile_id), 
+                        'company':Company.get_company(company_id=company_id)}
+            except (Company.DoesNotExist, CompanyAndUserRelation.DoesNotExist) as e:
+                return Response({'error': str(e)}, status=status.HTTP_404_NOT_FOUND)
+            serializer = SubscriptionSerializer(data=data)
+            if serializer.is_valid():
+                subscription = Subscription(**data)
+                subscription.save()
```

Comment:
Instead of manually creating and saving the `Subscription` instance, consider using `serializer.save()` after validating the serializer. This approach leverages DRF's serialization and validation mechanisms https://www.django-rest-framework.org/api-guide/serializers/

Reply:
Agree, fixed.